src/rootdev/rootdev.c

```diff
@@ -219,6 +219,82 @@ void rootdev_strip_partition(char *dst, size_t len) {
   *part = '\0';
 }
 
+/* Using stat does not work on btrfs because there is no 1-to-1 mapping
+ * to a single backing block device in general but we can get the first
+ * backing block device by looking in /proc/mounts
+ */
+#define PATH_SIZE 256
+dev_t rootdev_devt_from_mountpoint(const char *orig_path) {
+  FILE *mounts;
+  char path[PATH_SIZE];
+  char * line = NULL;
+  char * mountpoint = NULL;
+  char * devname = NULL;
+  size_t len = 0;
+  dev_t dev = 0;
+
+  mounts = fopen("/proc/mounts", "r");
+  if (mounts == NULL)
+        return 0;
+
+  strncpy(path, orig_path, PATH_SIZE - 1);
+  path[PATH_SIZE - 1] = '\0';
+  /* strip trailing "/" but not if it's the only character */
+  for (int i=strlen(path)-1; i>0; i--) {
+    if (path[i] == '/') {
+      path[i] = '\0';
+    } else {
+      break;
+    }
+  }
+
+  while (getline(&line, &len, mounts) != -1) {
+        devname = strtok(line, " ");
+        if (devname == NULL)
+          continue;
+        mountpoint = strtok(NULL, " ");
+        if (mountpoint == NULL)
+          continue;
+        if (strcmp(mountpoint, path) == 0) {
+          char devname_buf[PATH_SIZE];
+          char devname_parent[PATH_SIZE];
+          char sys_block_path[PATH_SIZE * 3];
+          char *tmp;
+
+          if (readlink(devname, (char *)&devname_buf, PATH_SIZE - 1) != -1) {
+            devname_buf[PATH_SIZE - 1] = '\0';
+            devname = devname_buf;
+          }
+          devname = strtok(devname, "/");
+          if (devname == NULL)
+            break;
+          /* skip to last part of the path */
+          while (tmp=strtok(NULL,"/"), tmp != NULL) {
+            devname = tmp;
+          }
+
+          strncpy(devname_parent, devname, PATH_SIZE);
```

Comment:
Elsewhere `PATH_SIZE - 1` is used with `strncpy()`, maybe do the same here to stay consistent.

Reply:
Yes, funny that GCC didn't complain as in cgpt